### PR TITLE
Feature/wsadapter (#1423)

### LINKED ISSUE #1423: Frontend Decouple Ws Adapter
### Description

Allow external web application to extend existing ws infrastructure

### Solution Proposed

_No response_

### Impact of Solution

_No response_

### Additional Context

_No response_

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.
- [ ] Check if a new demo could be provided based on this, or if legacy demos could be benefit from it.
- [ ] Ensure any change is well documented.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [X] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by dinhlongviolin1**
Resolve #1423

### CODE REVIEW

File: frontend/taipy-gui/base/src/packaging/taipy-gui-base.d.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    getRoutes(): Route[] | undefined;
+    update(encodedName: string, value: unknown): void;
+    getContext(): string;
+    updateContext(path?: string | undefined): void;
+    trigger(actionName: string, triggerId: string, payload?: Record<string, unknown>): void;
+    upload(encodedName: string, files: FileList, progressCallback: (val: number) => void): Promise<string>;
+    getPageMetadata(): any;
+}
+
+export {};
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/base/src/packaging/taipy-gui-base.d.ts]: is this needed ?

**Review by FredLL-Avaiga — DISMISSED** [review]
😶‍🌫️

File: frontend/taipy-gui/base/src/packaging/taipy-gui-base.d.ts (hunk 1 above)
Comment (dinhlongviolin1) [reply frontend/taipy-gui/base/src/packaging/taipy-gui-base.d.ts]: not at all! tx Fred

File: frontend/taipy-gui/base/src/packaging/taipy-gui-base.d.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    payload: Record<string, unknown> | unknown;
+    propagate: boolean;
+    client_id: string;
+    module_context: string;
+    ack_id?: string;
+}
+export declare const sendWsMessage: (socket: Socket | undefined, type: WsMessageType | str, name: string, payload: Record<string, unknown> | unknown, id: string, moduleContext?: string, propagate?: boolean, serverAck?: (val: unknown) => void) => string;
+export declare abstract class WsAdapter {
+    abstract supportedMessageTypes: string[];
+    abstract handleWsMessage(message: WsMessage, app: TaipyApp): void;
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/base/src/packaging/taipy-gui-base.d.ts]: I wonder if this method should return a Boolean value indicating that the message was processed, and that it should not be processed by further adapters. That allows for prioritizing message processing.
And that implies that the order of the registration is important: last registered, first served.

File: frontend/taipy-gui/base/src/packaging/taipy-gui-base.d.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+}
+export type OnInitHandler = (taipyApp: TaipyApp) => void;
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/base/src/packaging/taipy-gui-base.d.ts:23]: ... with respect to the previous comment, these would therefore return a Boolean value.

File: taipy/gui/gui.py [hunk 4]
```diff
@@ -652,6 +660,10 @@ def _manage_message(self, msg_type: _WsType, message: dict) -> None:
             else:
                 _warn(f"Decoding Message has failed: {message}", e)
 
+    # To be expanded by inheriting classes
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py:663]: Can you add a comment somewhere what the distinction 'internal' vs 'external' messages is?

**Review by FabienLelaquais — DISMISSED** [review]
Excellent!!

**Review by FredLL-Avaiga — DISMISSED** [review]
🎇

**Review by FabienLelaquais — DISMISSED** [review]
👌

**Review by FabienLelaquais — APPROVED** [review]
Great stuff!

**Pull request merged**